uboot-mediatek: support GL.iNet GL-X3000 and GL-XE3000

arachsys opened the pull request:
Add u-boot support based on the kernel dts introduced in d1016446 by @JeThWifirst and the GL-MT6000 u-boot support in fe10f974 from @zhaojh329 at GL.iNet.

The `pcie-mediatek-gen3` kernel driver doesn't like hotplug, so to work in PCIe mode, the 5G modem on this device needs to be switched on by u-boot before starting the kernel. Include an `init_modem` step in the `boot_system` action to set the relevant gpios. (The factory bootloader does the same, using Mediatek SDK-specific `gpio_power_clr` and `gpio_pull_up`.)

For PR reviewers: I'm not sure what the OpenWrt policy is with duplicating factory bootloader behaviour like this. I'd be fine to drop the `init_modem` as the OpenWrt scripts don't yet support PCIe MBIM and QMI, although this might cause problems when booting the factory kernel.

This has been tested on a real GL-X3000 device, both using `mtk_uartboot` with the OpenWrt-packaged `mt7981-ram-ddr4-bl2.bin`, and also by flashing `openwrt-mediatek-filogic-glinet_gl-x3000-preloader.bin` to `/dev/mmcblk0boot0` and `openwrt-mediatek-filogic-glinet_gl-x3000-bl31-uboot.fip` to `/dev/mmcblk0p4` (the fip partition).

For what it's worth, as well as OpenWrt, I'm able to boot a mainline 6.9.3 kernel fine on this device using this bootloader, although the block NVMEM and mac-base patch series are helpful for mac addresses and wifi eeprom.

zhaojh329 commented on package/boot/uboot-mediatek/patches/454-add-glinet-x3000.patch:
```diff
@@ -0,0 +1,290 @@
++	};
++};
++
++&eth {
++	status = "okay";
++	mediatek,gmac-id = <1>;
++	phy-mode = "gmii";
++	mediatek,switch = "mt7531";
++	reset-gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
++
++	fixed-link {
++		speed = <1000>;
++		full-duplex;
++	};
++};
```
These two devices do not use mt7531.
```
&eth {
        status = "okay";
        mediatek,gmac-id = <1>;
        phy-mode = "gmii";
        phy-handle = <&phy0>;

        mdio {
                phy0: ethernet-phy@0 {
                        compatible = "ethernet-phy-id03a2.9461";
                        reg = <0x0>;
                        phy-mode = "gmii";
                };
        };
};
```

arachsys replied:
Thanks! I've made this change in the device tree and checked tftp booting still works. I'll update the pull request now.

zhaojh329 commented on package/boot/uboot-mediatek/patches/454-add-glinet-x3000.patch:
```diff
@@ -0,0 +1,290 @@
++	status = "okay";
++};
++
++&watchdog {
++	status = "disabled";
++};
```
This is not required, watchdog is disabled by default in mt7981.dtsi.

arachsys replied:
Thanks!

mrkiko commented:
So nice!!
Thanks a lot for this work.
May I ask you to add some installation instructions?

mrkiko commented:
I mean - they are present in the PR description, but I can't see them in commit message.

arachsys commented:
> So nice!! Thanks a lot for this work. May I ask you to add some installation instructions?

Yes I agree, they'd be good to add to the commit message. I installed very 'manually' by opening the device up, then using a serial console and the u-boot mmc write command to the relevant partitions. You could probably write the bl31-uboot.fip to the fip partition directly from linux instead but probably not the preloader as the hardware boot partition will be write protected.

Presumably there's an official way to do this via the manufacturer's built in u-boot web server, and some of the standard images an openwrt build generates will do this if they're installed in the right way? I confess I'm not actually an openwrt user so don't really know what I should be recommending here. Happy to include the advice if someone can confirm what it should be!

(To explain: I actually built this gl-x3000 u-boot to run my own mainline kernel and aarch64 userspace which aren't openwrt-derived, but I'd like to contribute the bootloader back to openwrt, though, as it'll be much more widely useful to others here.)

mrkiko commented:
On Fri, Jun 07, 2024 at 06:34:23AM -0700, Chris Webb wrote:
> > So nice!! Thanks a lot for this work. May I ask you to add some installation instructions?
> 
> Yes I agree, they'd be good to add to the commit message. I installed very 'manually' by opening the device up, then using a serial console and the u-boot mmc write command to the relevant partitions. You could probably write the bl31-uboot.fip to the fip partition directly from linux instead but probably not the preloader as the hardware boot partition will be write protected.
> 
> Presumably there's an official way to do this via the manufacturer's built in u-boot web server, and some of the standard images an openwrt build generates will do this if they're installed in the right way? I confess I'm not actually an openwrt user so don't really know what I should be recommending here. Happy to include the advice if someone can confirm what it should be!
> 
> (To explain: I actually built this gl-x3000 u-boot to run my own mainline kernel and aarch64 userspace which aren't openwrt-derived, but I'd like to contribute the bootloader back to openwrt, though, as it'll be much more widely useful to others here.)

@dangowrt ?

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/15645#issuecomment-2154852119
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

JeThWifirst commented:
> Presumably there's an official way to do this via the manufacturer's built in u-boot web server

Indeed, as explained in the GL.iNet documentation: https://docs.gl-inet.com/router/en/4/faq/debrick/. 

Simply press and hold the reset button while powering the device, and after 5 seconds you can access the u-boot update page at http://192.168.1.1/uboot.html:
![image](https://github.com/openwrt/openwrt/assets/139239591/b0d1e18e-380e-408a-990d-80f905d9d946)

However I don't know which u-boot binary is expected by the updater!

arachsys commented:
> However I don't know which u-boot binary is expected by the updater!

Perhaps @zhaojh329 might be able to confirm with his GL.iNet hat on? :)

zhaojh329 commented:
> > However I don't know which u-boot binary is expected by the updater!
> 
> Perhaps @zhaojh329 might be able to confirm with his GL.iNet hat on? :)
> 
> My best guess based on older devices is that openwrt-mediatek-filogic-glinet_gl-x3000-bl31-uboot.fip is the right one. (This would imply you can only replace u-boot+bl31 not the bl2 preloader through the web interface.)

```
http://192.168.1.1/bl2.html
```

zhaojh329 commented:
> > Presumably there's an official way to do this via the manufacturer's built in u-boot web server
> 
> Indeed, as explained in the GL.iNet documentation: https://docs.gl-inet.com/router/en/4/faq/debrick/.
> 
> Simply press and hold the reset button while powering the device, and after 5 seconds you can access the u-boot update page at http://192.168.1.1/uboot.html: ![image](https://private-user-images.githubusercontent.com/139239591/337674785-b0d1e18e-380e-408a-990d-80f905d9d946.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3NzA4MDYsIm5iZiI6MTcxNzc3MDUwNiwicGF0aCI6Ii8xMzkyMzk1OTEvMzM3Njc0Nzg1LWIwZDFlMThlLTM4MGUtNDA4YS05OTBkLTgwZjkwNWQ5ZDk0Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwN1QxNDI4MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZTEzNWQ0ZWJmZDYwZmI1MDIwZTZhODY3NjNmMjg1NTNhNWRmYmEzMWExNzQ2YzczZGMwYzg2ZjAzYmY4YjQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.UjMRslVhuQnvZl687tNkVrT0v4h6KRaxNJgXCfdXqQ0)
> 
> However I don't know which u-boot binary is expected by the updater!

This page is for `openwrt-mediatek-filogic-glinet_gl-x3000-bl31-uboot.fip`.

arachsys commented:
As suggested, I've added brief instructions on using the web interface to upgrade bl2 and u-boot via the factory recovery mode.

arachsys committed:
uboot-mediatek: support GL.iNet GL-X3000 and GL-XE3000

Add u-boot support based on the kernel dts introduced in d1016446 and
the GL-MT6000 u-boot support in fe10f974.

The pcie-mediatek-gen3 kernel driver doesn't like hotplug, so to work in
PCIe mode, the 5G modem on this device needs to be switched on by u-boot
before starting the kernel. Include an init_modem step in the boot_system
action to set the relevant gpios. (The factory bootloader does the same,
using Mediatek SDK-specific gpio_power_clr and gpio_pull_up.)

Ideally the modem would be started using gpio-hog in the device tree, but
this will need to wait until mediatek gpio-hog support is fixed upstream:

  https://lore.kernel.org/u-boot/6ef2583e85eea60560d7776377d662779e7c44e5.1722419839.git.chris@arachsys.com/

The bootloader can be replaced using the built-in web interface of the
factory bootloader. Hold the reset button for five seconds while powering
on the device and it will boot into a recovery http server.

http://192.168.1.1/uboot.html and http://192.168.1.1/bl2.html can then
be used to upload openwrt-mediatek-filogic-glinet_gl-x3000-bl31-uboot.fip
and openwrt-mediatek-filogic-glinet_gl-x3000-preloader.bin respectively.

Alternatively, from a root shell on the running system, unlock the boot
partition with

  echo 0 >/sys/block/mmcblk0boot0/force_ro

then write openwrt-mediatek-filogic-glinet_gl-x3000-bl31-uboot.fip to
/dev/mmcblk0p4 and openwrt-mediatek-filogic-glinet_gl-x3000-preloader.bin
to /dev/mmcblk0boot0.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys commented:
I wonder if it would be possible to initialise the gpios using gpio-hog in the u-boot device tree instead of an init_modem action? I wondered about something like
```
&gpio {
	hub_power {
		gpio-hog;
		gpios = <5 GPIO_ACTIVE_HIGH>;
		output-high;
	}

	modem_control {
		gpio-hog;
		gpios = <9 GPIO_ACTIVE_HIGH>;
		output-high;
	}

	modem_reset {
		gpio-hog;
		gpios = <10 GPIO_ACTIVE_HIGH>;
		output-low;
	}

	modem_power {
		gpio-hog;
		gpios = <11 GPIO_ACTIVE_HIGH>;
		output-high;
	}
}
```
but as far as I can see, this isn't sufficient. Can someone more experienced with device tree and u-boot advise/comment?

mrkiko commented:
Hopng this PR can progress.

arachsys commented:
Hi, I'm just about to post a corrected patch for the u-boot gpiohog issue on the u-boot mailing list. Once that is merged, I can backport as part of this PR, and then we can use standard gpiohog in the device tree for powering up the modem instead of the special init_modem action in the u-boot environment. I'll update this PR once u-boot has merged the mix.

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -643,14 +643,19 @@ define Device/glinet_gl-x3000-xe3000-common
     kmod-fs-f2fs kmod-hwmon-pwmfan kmod-usb3 kmod-usb-serial-option \
     kmod-usb-storage kmod-usb-net-qmi-wwan uqmi
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to 32M | append-rootfs
```
Why do you add a factory image? Is this needed for OpenWrt?

arachsys replied:
Cargo cult I'm afraid, to match what was done when adding the GL-MT6000 (Flint 2) u-boot support.

hauke replied:
ok, this should not be needed for the u-boot. Maybe the vendor bootloader of the GL-MT6000  needs a special image. It is probably better to leave it our in this PR.

hauke replied:
ok, @mrkiko  thinks it is needed, maybe leave it in.

arachsys replied:
Great okay, as requested it's still there in the rebased version I pushed.

hauke commented:
Please rebase the PR. github shows that it has some conflicts.

Thanks for getting your change upstream. I think this PR is already fine. You can send an update PR when you got it accepted upstream.

mrkiko commented:
Anyone feel free to correct me - but I guess you need a factory image because contrarily to gl's uboot, upstream one doesn't handle sysupgrade image formats.
Is this correct? Infact a factory image was added also for the mt6000, to be used only if the user decised to switch to upstream u-boot.

On Fri, Aug 30, 2024 at 06:19:24AM -0700, Hauke Mehrtens wrote:
> @hauke commented on this pull request.
> 
> 
> 
> > @@ -643,14 +643,19 @@ define Device/glinet_gl-x3000-xe3000-common
>    DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware mkf2fs \
>      kmod-fs-f2fs kmod-hwmon-pwmfan kmod-usb3 kmod-usb-serial-option \
>      kmod-usb-storage kmod-usb-net-qmi-wwan uqmi
> +  IMAGES += factory.bin
> +  IMAGE/factory.bin := append-kernel | pad-to 32M | append-rootfs
> 
> Why do you add a factory image? Is this needed for OpenWrt?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/15645#pullrequestreview-2272313055
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

arachsys commented:
> Please rebase the PR. github shows that it has some conflicts.
> 
> Thanks for getting your change upstream. I think this PR is already fine. You can send an update PR when you got it accepted upstream.

Great okay! I've rebased this version with the init script onto current openwrt main.

hauke commented:
Applied to main branch in 7eda44f53647cb54cb9b2eb13b27417cd0128a37.
